control/controlhttp/client.go

```diff
@@ -100,14 +87,26 @@ func (a *dialParams) dial(ctx context.Context) (*controlbase.Conn, error) {
 	// u80 and u443 are the URLs we'll try to hit over HTTP or HTTPS,
 	// respectively, in order to do the HTTP upgrade to a net.Conn over which
 	// we'll speak Noise.
+	var port80, port443 string
+	if a.HTTPPort != "" {
+		port80 = a.HTTPPort
+	} else {
+		port80 = "80"
+	}
+	if a.HTTPSPort != "" {
+		port443 = a.HTTPSPort
+	} else {
+		port443 = "443"
+	}
+
 	u80 := &url.URL{
 		Scheme: "http",
-		Host:   net.JoinHostPort(a.host, a.httpPort),
+		Host:   net.JoinHostPort(a.Hostname, port80),
```

Comment:
or some helper and like `net.JoinHostPort(a.Hostname, strDef(a.HTTPPort, "80"))` here and line 109